crates/uv-resolver/src/marker.rs

```diff
@@ -226,6 +247,177 @@ pub(crate) fn normalize(tree: MarkerTree) -> Option<MarkerTree> {
     }
 }
 
+/// Removes redundant expressions from the tree recursively.
+///
+/// This function does not attempt to flatten or clean the tree and may leave it in a denormalized state.
+pub(crate) fn filter_all(tree: &mut MarkerTree) {
+    match tree {
+        MarkerTree::And(trees) => {
+            for subtree in &mut *trees {
+                filter_all(subtree);
+            }
+
+            for conjunct in collect_expressions(trees) {
+                // Filter out redundant disjunctions (by the Absorption Law).
+                trees.retain_mut(|tree| !filter_disjunctions(tree, &conjunct));
+
+                // Filter out redundant expressions in this conjunction.
+                for tree in &mut *trees {
+                    filter_conjunct_exprs(tree, &conjunct);
+                }
+            }
+        }
+
+        MarkerTree::Or(trees) => {
+            for subtree in &mut *trees {
+                filter_all(subtree);
+            }
+
+            for disjunct in collect_expressions(trees) {
+                // Filter out redundant conjunctions (by the Absorption Law).
+                trees.retain_mut(|tree| !filter_conjunctions(tree, &disjunct));
+
+                // Filter out redundant expressions in this disjunction.
+                for tree in &mut *trees {
+                    filter_disjunct_exprs(tree, &disjunct);
+                }
+            }
+        }
+
+        MarkerTree::Expression(_) => {}
+    }
+}
+
+// Collect all expressions from a tree.
+fn collect_expressions(trees: &[MarkerTree]) -> Vec<MarkerExpression> {
+    trees
+        .iter()
+        .filter_map(|tree| match tree {
+            MarkerTree::Expression(expr) => Some(expr.clone()),
+            _ => None,
+        })
+        .collect()
+}
+
+// Filters out matching expressions from any nested disjunctions.
+fn filter_disjunct_exprs(tree: &mut MarkerTree, disjunct: &MarkerExpression) {
+    match tree {
+        MarkerTree::Or(trees) => {
+            trees.retain_mut(|tree| match tree {
+                MarkerTree::Expression(expr) => expr != disjunct,
+                _ => {
+                    filter_disjunct_exprs(tree, disjunct);
+                    true
+                }
+            });
+        }
+        MarkerTree::And(trees) => {
+            for tree in trees {
+                filter_disjunct_exprs(tree, disjunct);
+            }
+        }
+
+        MarkerTree::Expression(_) => {}
+    }
+}
+
+// Filters out matching expressions from any nested conjunctions.
```

Comment:
Is there a reason why this doesn't also filter them out for a direct leaf expression? I guess it would need to rewrite the marker expression as `MarkerTree::And(vec![])` in that case...

Another perspective that I find helpful for routines like this is phrasing like, "returns a marker tree in which the given expression is assumed to be true in all conjunctions." I think that captures better the essence of the contract here if I'm understanding correctly.

Reply:
Yeah, that's covered by the `dedup` call in `normalize_all`. It's a bit trickier to do that here.